base/rational.jl

```diff
@@ -95,6 +95,22 @@ end
 function write(s::IO, z::Rational)
     write(s,numerator(z),denominator(z))
 end
+function parse(::Type{Rational{T}}, s::AbstractString) where T<:Integer  
+    ss = split(s, '/'; limit = 2)
+    if isone(length(ss))
+        return parse(T, only(ss)) // one(T) 
+    end    
+    @assert length(ss) == 2
```

Comment:
```suggestion
    isempty(ss) && throw(ArgumentError("input string is empty or only contains whitespace"))
```
`@assert` should not be used to sanitize user input, since there is no guarantee on the check actually occurring. In this particular case, it can also useful to provide a clear error message since the only wrong case is for the empty input.

I propose the error message of `parse(Int, "")` for consistency.

Reply:
`@assert` is not being used to sanitize input. The `limit` keyword of `split` guarantees that the length is either 1 or 2. The length can never be zero, because if an empty string is passed, then `ss` will be an array of **one** element with an empty string inside. The current error is already `parse(Int, "")` (if `T` == `Int`) because vector with an empty single string will enter the `if` and be parsed as a single value. The `@assert` purpose is to indicate to the compiler and the programmer an invariant, in other words, the correct use of `@assert`.

Reply:
Ah yes you're absolutely right! Sorry about that.
However, the compiler won't pick on the indication that the size is 2 because there is no modeling of the size of a vector in the compiler, so you should just get rid of the `@assert` and replace with a comment for example. If you *really* want to give that indication to the compiler, you can write `ns = (@inbounds ss[1]); ds = (@inbounds ss[2])` below, but I wouldn't actually do it here.

Just a note, to see that the information from `@assert` is not passed to the compiler:
```julia
julia> function foo(x)
           @assert length(x) == 2
           a, b = x
           return b
       end;

julia> @code_typed foo([1,2])
CodeInfo(
1 ─ %1 = Base.arraylen(x)::Int64
│   %2 = (%1 === 2)::Bool
└──      goto #3 if not %2
2 ─      nothing::Nothing
│        Base.arrayref(true, x, 1)::Int64
│   %6 = Base.arrayref(true, x, 2)::Int64
└──      return %6
3 ─ %8 = %new(Core.AssertionError, "length(x) == 2")::AssertionError
│        Base.throw(%8)::Union{}
└──      unreachable
) => Int64
```
you can see that the `arrayref` calls start by a `true` argument, which indicates that there will be bounds checking regardless of the assert.

Reply:
No need to be sorry. I should have used `@code_typed` as you pointed out, I added the `@assert` without knowing it would be effective or not. I checked and `@inbounds ns, ds = ss[1], ss[2]` drops the check, and it is not so ugly, do you think I can incorporate it? As the `parse` can be heavily called when bringing data from a file, then maybe it is good for it to be optimized where possible.

Reply:
I don't have a strong opinion either way, I'd say it's fine to incorporate it if you want.

Reply:
Will do it then.